tlair/src/main/scala/at/forsyte/apalache/tla/lir/TlaDeclLevelFinder.scala

```diff
@@ -0,0 +1,55 @@
+package at.forsyte.apalache.tla.lir
+
+/**
+ * This class computes the level of a TLA+ declaration by using TlaExLevelFinder. See Lamport. Specifying Systems, p. 322.
+ *
+ * @param module a TLA module that contains the declarations to query
+ * @author Igor Konnov
+ */
+class TlaDeclLevelFinder(module: TlaModule) {
+  private val consts = Set(module.constDeclarations.map(_.name): _*)
+  private val vars = Set(module.varDeclarations.map(_.name): _*)
+  private val defs = Map(module.operDeclarations map { d => d.name -> d }: _*)
+  private var levelCacheMap: Map[String, TlaLevel] = Map()
+
+  def apply(decl: TlaDecl): TlaLevel = {
+    decl match {
+      case TlaConstDecl(_) =>
+        TlaLevelConst
+
+      case TlaVarDecl(_) =>
+        TlaLevelState
+
+      case TlaAssumeDecl(_) =>
+        TlaLevelConst
+
```

Comment:
unless I missunderstand, it seems like where checking for this and then determining the level in two parallel ways, Once here by pattern matching, and once when we construct the level finder, when we cache to the various sets. Couldn't we handle all of this logic with the `_.contains` checks and the caching lookup?

Reply:
The code in `apply` is also handling the cases of `ASSUME` and `THEOREM`. Also, this code avoids an extra lookup in `defs`, `consts` and `vars` (but that's not the main point).

Reply:
What happens if I write an assume declaration with primes?

Reply:
SANY will tell you that it does not like your spec. I found that out a couple of months ago :D

Reply:
It will even complain if you use state variables in `ASSUME`.

Reply:
Ok, what about the expression `(1 + 1)'` outside of assume? Just because it has a prime operator, does not mean its level isn't 0.